espresso/environment/2_espresso_liveness_test.go

```diff
@@ -119,3 +131,285 @@ func TestE2eDevNetWithEspressoEspressoDegradedLiveness(t *testing.T) {
+// transaction should be less than 10 seconds.
+//
+// More importantly, this **SHOULD** also continue to be the state even when
+// Espresso is in a degraded state.
+//
+// Sadly, there does not seem to be an easy way to associate the Transaction
+// submitted to the L2 with the Block being returned from Espresso.  However,
+// in this test scenario, we know that the Batch number will correspond to the
+// transaction number we submitted to the sequencer.  As a result, we should
+// be able to match the Batch number to the transaction number in transaction
+// order.
+
+func TestE2eDevNetWithEspressoEspressoDegradedLivenessViaCaffNode(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+	defer cancel()
+
+	launcher := new(env.EspressoDevNodeLauncherDocker)
+
+	// Start a Server to proxy requests to Espresso, with a decider that will
+	// simulate degraded liveness failures by reporting false successful
+	// submissions 10% of the time, and 503 errors 10% of the time, with
+	// actual proxied requests 80% of the time.
+	_, server, option := env.SetupQueryServiceIntercept(
+		env.SetDecider(env.NewRandomRollFakeSubmitTransactionSuccess(
+			10,
+			0,
+			1,
+			rand.New(rand.NewSource(0)),
+		)),
+	)
+
+	defer env.Stop(t, server)
+	system, espressoDevNode, err := launcher.StartDevNet(ctx, t, 0, option)
+
+	// Signal the testnet to shut down
+	if have, want := err, error(nil); have != want {
+		t.Fatalf("failed to start dev environment with espresso dev node:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+	}
+
+	defer env.Stop(t, system)
+	defer env.Stop(t, espressoDevNode)
+
+	caffNode, err := env.LaunchDecaffNode(t, system, espressoDevNode)
+	if have, want := err, error(nil); have != want {
+		t.Fatalf("failed to start caff node:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+	}
+
+	// Shut down the Caff Node
+	defer env.Stop(t, caffNode)
+
+	addressAlice := system.Cfg.Secrets.Addresses().Alice
+
+	l1Client := system.NodeClient(e2esys.RoleL1)
+	l2Seq := system.NodeClient(e2esys.RoleSeq)
+	caffVerif := system.NodeClient(env.RoleCaffNode)
+
+	balanceAliceInitial, err := caffVerif.BalanceAt(ctx, addressAlice, nil)
+	if have, want := err, error(nil); have != want {
+		t.Fatalf("Failed to fetch Alice's balance:\nhave:\n\t\"%v\"\nwant:\n\t\"%v\"\n", have, want)
+	}
+
+	type espressoReceived struct {
+		batch    *derive.EspressoBatch
+		block    *geth_types.Block
+		received time.Time
+	}
+
+	espressoReceipts := map[common.Hash]espressoReceived{}
+
+	streamBlocksCtx, streamBlocksCancel := context.WithCancel(ctx)
+	var wg sync.WaitGroup
+	defer streamBlocksCancel()
+	{
+		// Streamer Setup and Configuration
+		l := log.NewLogger(slog.Default().Handler())
+		streamer := espresso.NewEspressoStreamer(
+			system.RollupConfig.L2ChainID.Uint64(),
+			batcher.NewAdaptL1BlockRefClient(l1Client),
+			espressoClient.NewClient(server.URL),
+			nil, // TODO(AG)
```

Comment:
Maybe no need to copy the TODO here.

Reply:
Addressed in cdf5e56d23f7ae4b5ff688378e345e6d856e162f.